### PR TITLE
Refactor `AuthorizeSiteAccess`

### PR DESCRIPTION
**Pull request by zoldar**
### Changes

This PR is refactoring `AuthorizeSiteAccess`, adding some extra compile- and runtime checks along the way. The range of accepted inputs was revised as well.

### Tests
- [x] Automated tests have been added

### CODE REVIEW

File: lib/plausible_web/plugs/authorize_site_access.ex [hunk 1]
```diff
@@ -58,21 +60,51 @@ defmodule PlausibleWeb.AuthorizeSiteAccess do
...
+    end
+  end
+
+  defp maybe_get_shared_link(conn, site) do
+    slug = conn.path_params["slug"] || conn.params["auth"]
+    site_id = site.id
+
+    if is_binary(slug) do
+      case Repo.get_by(Plausible.Site.SharedLink, slug: slug) do
+        %{site_id: ^site_id} = shared_link -> {:ok, shared_link}
```
Comment (ukutaht) [inline lib/plausible_web/plugs/authorize_site_access.ex:84]: Can the `site_id == ^site_id` condition be applied in the SQL query itself? i.e.

```elixir
if link = Repo.get_by(Plausible.Site.SharedLink, slug: slug, site_id: site.id) do
  {:ok, link}
else
  error_not_found(conn)
end
```

**Review by ukutaht — COMMENTED** [review]
Opportunity to remove another DB roundtrip: https://github.com/plausible/analytics/commit/9ed96632adf44c211d462ac4b144514634d1720f

I think the code in my proposal is less elegant than the PR. But since auth code runs for pretty much every controller action, I like taking some extra care with performance there.

File: lib/plausible_web/plugs/authorize_site_access.ex (hunk 1 above)
Comment (zoldar) [reply lib/plausible_web/plugs/authorize_site_access.ex]: Thanks, fixed in https://github.com/plausible/analytics/pull/4583/commits/9db87fc2677a9acf54aaad08580f80c52205c9ab. Not sure what was going through my head when writing this...

**Comment by zoldar** [comment]
> Opportunity to remove another DB roundtrip: https://github.com/plausible/analytics/commit/9ed96632adf44c211d462ac4b144514634d1720f
>
> I think the code in my proposal is less elegant than the PR. But since auth code runs for pretty much every controller action, I like taking some extra care with performance there.

Agreeed. Applied your change with a minor adjustment of naming.

**Review by ukutaht — APPROVED** [review]

**Pull request merged**